File: app/XtalPackage/Sources/XtalFeature/Autocomplete/Fantasma.swift
import AppKit

/// El `NSTextView` del editor, con una sola cosa agregada: sabe dibujar el **fantasma**,
/// el texto gris que el autocomplete propone a la derecha del cursor.
///
/// ## Por qué se dibuja y no se inserta
///
/// La tentación es meter la sugerencia en el texto con un color gris y sacarla si la
/// persona sigue escribiendo. Es una mala idea por tres razones concretas:
///
/// 1. **El archivo se guarda en cada tecla.** Con el texto insertado, el fantasma se
///    guardaría en el `.tex`. Alcanza con que la app se cierre en el momento justo.
/// 2. **Rompe el undo.** Cada aparición y cada borrado del fantasma entraría al historial,
///    y ⌘Z dejaría de deshacer lo que uno escribió.
/// 3. **Ensucia todo lo que lee el texto**: el coloreado, la sincronía con el PDF y el
///    propio autocomplete, que terminaría leyéndose a sí mismo.
///
/// Dibujarlo encima no toca el documento. Lo que se ve es una pintura, y hasta que alguien
/// apriete Tab no existe en ningún lado.
///
/// ## Cómo encuentra dónde pintar
///
/// Con `firstRect(forCharacterRange:)`, la misma que usa la lista de `\omega` para pararse
/// debajo del renglón. Viene en coordenadas de **pantalla**, así que hay que traerla dos
/// veces: a la ventana y de ahí a la vista.
///
/// Contraparte: `app-win/src/editor/fantasma.ts` (un decorador de CodeMirror).
final class VistaEditor: NSTextView {
    /// Lo que se pinta en gris. `nil` = no hay nada.
    var fantasma: String? {
        didSet {
            guard fantasma != oldValue else { return }
            needsDisplay = true
        }
    }

    /// En qué posición del texto vale. Si el cursor se movió, no se pinta: mostrarlo donde
    /// está el cursor ahora sería mostrar una sugerencia hecha para otro lugar.
    var posicionDelFantasma: Int = -1 {
        didSet { needsDisplay = true }
    }

    override func draw(_ dirtyRect: NSRect) {
        super.draw(dirtyRect)

        guard let fantasma, !fantasma.isEmpty else { return }
        let cursor = selectedRange()
        guard cursor.length == 0, cursor.location == posicionDelFantasma else { return }
        guard let ventana = window else { return }

        let enPantalla = firstRect(forCharacterRange: NSRange(location: cursor.location, length: 0),
                                   actualRange: nil)
        guard enPantalla.height > 0 else { return }
        let enVentana = ventana.convertFromScreen(enPantalla)
        let local = convert(enVentana, from: nil)

        let tipografia = font ?? .monospacedSystemFont(ofSize: 12.5, weight: .regular)
        let atributos: [NSAttributedString.Key: Any] = [
            .font: tipografia,
            // Gris apagado, y no el color del texto con transparencia: sobre un fondo
            // oscuro la transparencia lo deja casi blanco y deja de leerse como propuesta.
            .foregroundColor: NSColor.tertiaryLabelColor,
        ]

        // La caja va del cursor al borde derecho, con alto para las líneas que como mucho
        // puede tener la sugerencia. `draw(in:)` respeta los saltos de línea; `draw(at:)`
        // los ignora y sale todo pisado en un renglón.
        let alto = tipografia.boundingRectForFont.height
        let caja = NSRect(x: local.minX,
                          y: local.minY,
                          width: max(0, bounds.width - local.minX - textContainerInset.width),
                          height: alto * 4)
        (fantasma as NSString).draw(in: caja, withAttributes: atributos)
    }

    /// Se llama desde el editor cada vez que el control cambia de opinión.
    func mostrar(_ texto: String?, en posicion: Int) {
        posicionDelFantasma = posicion
        fantasma = texto
    }
}

File: app-win/src/editor/fantasma.ts
/**
 * El **fantasma**: el texto gris que el autocomplete propone a la derecha del cursor.
 *
 * Contraparte de `app/…/Autocomplete/Fantasma.swift`. Lo que allá es dibujar encima de un
 * `NSTextView`, acá es un widget de CodeMirror.
 *
 * ## Por qué un widget y no texto insertado
 *
 * La tentación es meter la sugerencia en el documento con un color gris y sacarla si la
 * persona sigue escribiendo. Es una mala idea por tres razones concretas:
 *
 * 1. **El editor guarda en cada tecla.** Con el texto insertado, el fantasma se guardaría
 *    en el `.tex`. Alcanza con que la app se cierre en el momento justo.
 * 2. **Rompe el undo.** Cada aparición y cada borrado entraría al historial, y Ctrl+Z
 *    dejaría de deshacer lo que uno escribió.
 * 3. **Ensucia todo lo que lee el documento**: el coloreado, la sincronía con el PDF y el
 *    propio autocomplete, que terminaría leyéndose a sí mismo.
 *
 * Un widget es DOM que CodeMirror dibuja adentro del editor **sin que exista en el
 * documento**: `state.doc` no lo ve, y por lo tanto no lo ve nada de lo de arriba.
 */

import { EditorView, Decoration, WidgetType, type DecorationSet } from "@codemirror/view";
import { StateEffect, StateField } from "@codemirror/state";

/** Poner o sacar el fantasma. `null` lo saca. */
export const ponerFantasma = StateEffect.define<{ texto: string; pos: number } | null>();

class WidgetFantasma extends WidgetType {
  constructor(readonly texto: string) {
    super();
  }

  /** Sin esto, CodeMirror rehace el DOM en cada actualización y el texto parpadea. */
  eq(otro: WidgetFantasma) {
    return otro.texto === this.texto;
  }

  toDOM() {
    const span = document.createElement("span");
    span.className = "cm-fantasma";
    span.textContent = this.texto;
    // El fantasma no es texto de verdad: no se puede seleccionar ni copiar, y el lector
    // de pantalla no tiene por qué leerlo como si ya estuviera escrito.
    span.setAttribute("aria-hidden", "true");
    return span;
  }

  /** No come clicks: tocar ahí tiene que poner el cursor donde uno tocó. */
  ignoreEvent() {
    return true;
  }
}

/**
 * Dónde vive la sugerencia dentro del estado del editor.
 *
 * Es un `StateField` y no una variable suelta porque CodeMirror **rehace las decoraciones
 * en cada cambio del documento**: una variable de afuera se perdería en la primera tecla,
 * y el campo se actualiza solo cuando el documento se mueve.
 */
export const campoFantasma = StateField.define<DecorationSet>({
  create() {
    return Decoration.none;
  },
  update(valor, tr) {
    // Cualquier cambio en el documento invalida la sugerencia: fue calculada para el
    // texto de antes.
    if (tr.docChanged) valor = Decoration.none;
    else valor = valor.map(tr.changes);

    for (const efecto of tr.effects) {
      if (!efecto.is(ponerFantasma)) continue;
      const v = efecto.value;
      if (!v || !v.texto) {
        valor = Decoration.none;
      } else {
        valor = Decoration.set([
          Decoration.widget({ widget: new WidgetFantasma(v.texto), side: 1 }).range(v.pos),
        ]);
      }
    }
    return valor;
  },
  provide: (f) => EditorView.decorations.from(f),
});

/** Si hay algo en pantalla. Lo mira el atajo de Tab para decidir si se come la tecla. */
export function hayFantasma(vista: EditorView): boolean {
  return vista.state.field(campoFantasma, false)?.size ? true : false;
}

export function sacarFantasma(vista: EditorView) {
  if (hayFantasma(vista)) vista.dispatch({ effects: ponerFantasma.of(null) });
}

/**
 * El color.
 *
 * Gris apagado, y **no el color del texto con transparencia**: sobre el fondo oscuro la
 * transparencia lo deja casi blanco y deja de leerse como una propuesta.
 */
export const temaFantasma = EditorView.theme({
  ".cm-fantasma": {
    color: "var(--texto-3)",
    opacity: "0.85",
    // El salto de línea de una sugerencia de dos renglones tiene que verse.
    whiteSpace: "pre-wrap",
  },
});
